contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py

```diff
@@ -0,0 +1,653 @@
+
+
+def tokenized_targets(
+    train_set: Dataset,
+    validation_set: Dataset,
+    inference_set: Dataset,
+    tokenizer: PreTrainedTokenizerBase,
+) -> DatasetType:
+    return concatenate_datasets([train_set, validation_set, inference_set]).map(
+        lambda x: tokenizer(x["output_text"], truncation=True),
+        batched=True,
+        remove_columns=["input_text", "output_text"],
+    )
+
+
+def target_lengths(tokenized_targets: DatasetType) -> list[int]:
+    """The lengths of the target sequences."""
+    return [len(x) for x in tokenized_targets["input_ids"]]
+
+
+def max_target_lengths(target_lengths: list[int]) -> int:
+    """The 95th percentile of the target lengths."""
+    return int(np.percentile(target_lengths, 95))
+
+
+def _preprocess_function(
+    sample,
+    max_source_lengths: int,
+    max_target_lengths: int,
+    tokenizer: PreTrainedTokenizerBase,
+    padding="max_length",
+):
+    """Helper function to preprocess the data."""
+    # add prefix to the input for t5
+    inputs = [item for item in sample["input_text"]]
+
+    # tokenize inputs
+    model_inputs = tokenizer(
+        inputs, max_length=max_source_lengths, padding=padding, truncation=True
+    )
+
+    # Tokenize targets with the `text_target` keyword argument
+    labels = tokenizer(
+        text_target=sample["output_text"],
+        max_length=max_target_lengths,
+        padding=padding,
+        truncation=True,
+    )
+
+    # If we are padding here, replace all tokenizer.pad_token_id in the labels by -100 when we want to ignore
+    # padding in the loss.
+    if padding == "max_length":
+        labels["input_ids"] = [
+            [(token if token != tokenizer.pad_token_id else -100) for token in label]
+            for label in labels["input_ids"]
+        ]
+
+    model_inputs["labels"] = labels["input_ids"]
+    return model_inputs
+
+
+def tokenized_train(
+    train_set: Dataset,
+    max_source_lengths: int,
+    max_target_lengths: int,
+    tokenizer: PreTrainedTokenizerBase,
+) -> Dataset:
+    """Tokenizes the training set."""
+    return train_set.map(
+        partial(
+            _preprocess_function,
+            max_source_lengths=max_source_lengths,
+            max_target_lengths=max_target_lengths,
+            tokenizer=tokenizer,
+        ),
+        batched=True,
+        remove_columns=["input_text", "output_text"],
+    )
+
+
+def tokenized_validation(
+    validation_set: Dataset,
+    max_source_lengths: int,
+    max_target_lengths: int,
+    tokenizer: PreTrainedTokenizerBase,
+) -> Dataset:
+    """Tokenizes the validation set."""
+    return validation_set.map(
+        partial(
+            _preprocess_function,
+            max_source_lengths=max_source_lengths,
+            max_target_lengths=max_target_lengths,
+            tokenizer=tokenizer,
+        ),
+        batched=True,
+        remove_columns=["input_text", "output_text"],
+    )
+
+
+def tokenized_inference(
```

Comment:
Having these all the same is a parallel pipeline -- maybe best to first tokenize and then split into datasets. That way part of this is reusable for inference if you want to infer in batch, and we don't repeat the same exact things three times. I'm assuming its OK to do this as its a pure map operation.

Reply:
the use case is different as the second tokenization is applied differently.